### DISCUSSION TITLE
How to use `nextLink` and `deltaLink`

### DISCUSSION THREAD

**Discussion by maxklav** [microsoftgraph/msgraph-sdk-go#533] [Q&A] [answered]
I'm able to retrieve them with the SDK calling `.Delta()` before the `.Get()`, but how do I use them?

```go
import (
	"context"

	msgraphsdk "github.com/microsoftgraph/msgraph-sdk-go"
	"github.com/microsoftgraph/msgraph-sdk-go/users"
)

func main() {
	ctx := context.Background()

	tokenCredential := "secret!"

	client, err := msgraphsdk.NewGraphServiceClientWithCredentials(tokenCredential, nil)
	if err != nil {
		return nil, err
	}

	params := &users.ItemMailFoldersDeltaRequestBuilderGetRequestConfiguration{
		QueryParameters: &users.ItemMailFoldersDeltaRequestBuilderGetQueryParameters{
			// nextLink somewhere here ?
		},
	}

	resp, err := client.Users().ByUserId(userID).MailFolders().Delta().Get(ctx, params)
	if err != nil {
		return nil, err
	}

	// How to use them
	nextLink := resp.GetOdataNextLink()
	deltaLink := resp.GetOdataDeltaLink()
}
```

How do we use the `nextLink` and `deltaLink` ?  
**Do I need to build a HTTP request with these links or can I just use the token from the link in the SDK ?**  
Is there code in the SDK that we can use to extract the `skiptoken` from the link ?  
How can I use this token in `client.Users().ByUserId(userID).MailFolders().Delta().Get(ctx, params)` ?  
Is it in the `params` ?  

My ultimate goal is to use `nextLink` until I have a `deltaLink` and save the `deltaLink` for later use.

**Comment by baywet** [comment]
Thanks for using the Go SDK and for reaching out. The Go SDK now provides a [page iterator](https://github.com/microsoftgraph/msgraph-sdk-go-core/blob/main/page_iterator.go) that'll go through the next links automatically for you. Once your application has enumerated all the current changes, it needs to store the delta link that'll be provided and call the API again at a later time.

Unfortunately we haven't had time yet to document the use of the page iterator in [our public docs](https://learn.microsoft.com/en-us/graph/sdks/paging?tabs=csharp) (CC @jasonjoh ) but you can look at other languages for the time being to get a global understanding. (And maybe @rkodev can provide a snippet here).

You should also have a look at the [delta queries concept page](https://learn.microsoft.com/en-us/graph/delta-query-overview) to get a better understanding of how delta queries work in general on Microsoft Graph.

**Reply by maxklav** [reply]
Thanks for your answer. I see in [page iterator code](https://github.com/microsoftgraph/msgraph-sdk-go-core/blob/main/page_iterator.go) the method `GetOdataNextLink()` but how to I use this link after with the iterator?

**Reply by maxklav** [reply]
I asked the question for `nextLink` and `deltaLink`, I'm actually already using an iterator and I want to know how to use the `deltaLink`:

```go
import (
	"context"

	msgraphsdk "github.com/microsoftgraph/msgraph-sdk-go"
)

func main() {
	ctx := context.Background()

	tokenCredential := "secret!"

	client, err := msgraphsdk.NewGraphServiceClientWithCredentials(tokenCredential, nil)
	if err != nil {
		fmt.Printf("error from client creation: %v\n", err)
		return
	}

	config := &sites.SitesRequestBuilderGetRequestConfiguration{
		QueryParameters: &sites.SitesRequestBuilderGetQueryParameters{
			Orderby: []string{"createdDateTime desc"},
		},
	}

	deltaLink := "previous-delta-link" // Where do I use/plug this one! 

	resp, err := me.graphSdkClient.Sites().Get(ctx, config)
	if err != nil {
		fmt.Printf("error from get sites: %v\n", err)
		return
	}

	pageIterator, err := msgraphcore.NewPageIterator[models.Siteable](
		resp,
		me.graphSdkClient.GetAdapter(),
		models.CreateSiteCollectionResponseFromDiscriminatorValue,
	)
	if err != nil {
		fmt.Printf("error from page iterator creation %v\n", err)
		return
	}

        if err := pageIterator.Iterate(ctx, func(pageItem models.Siteable) bool {
	       fmt.Printf("Site found: %v\n",  pageItem.GetWebUrl())
        
	        return true
        }); err != nil {
		fmt.Printf("error from iterator %v\n", err)
		return
        }

        // Save pageIterator.GetOdataDeltaLink() in DB for next time
}
```

I just don't know how to use the **delta link** (or delta token from the **delta link**) after, with or without iterator.

**Reply by baywet** [reply]
here https://github.com/microsoftgraph/msgraph-sdk-go/blob/fe2338883b7c7eafc484c1723416231a83a48e01/sites/site_item_request_builder.go#L53

**Reply by maxklav** [reply]
Thank you!

I'm currently using it for mail folders of a specific user:

```go
import (
	"context"

	msgraphsdk "github.com/microsoftgraph/msgraph-sdk-go"
)

func main() {
	ctx := context.Background()

	tokenCredential := "secret!"

	client, err := msgraphsdk.NewGraphServiceClientWithCredentials(tokenCredential, nil)
	if err != nil {
		// Handle error
		return
	}

	requestIncludehiddenfolders := "true"
	params := &users.ItemMailFoldersRequestBuilderGetRequestConfiguration{
		QueryParameters: &users.ItemMailFoldersRequestBuilderGetQueryParameters{
			IncludeHiddenFolders: &requestIncludehiddenfolders,
		},
	}
	adapter := client.GetAdapter()

	userID = "user-id"
	req := client.Users().ByUserId(userID).MailFolders()

	resp, err := req.Get(ctx, params)
	if err != nil {
		// Handle err
		return
	}

	for {
		iterator, err := createIterator(resp)
		if err != nil {
			// handle err
			return
		}

		if err := iterate(iterator); err != nil {
			// handle err
			return
		}

		// This is always nil
		deltaLink := iterator.GetOdataDeltaLink()
		if deltaLink == nil {
			// handle err
			return
		}

		time.Sleep(5 * time.Minute)

		req = users.NewItemMailFoldersRequestBuilder(*deltaLink, adapter)

		resp, err = req.Get(ctx, params)
		if err != nil {
			// handle err
			return
		}
	}
}


func createIterator(resp models.MailFolderCollectionResponseable) (*msgraphcore.PageIterator[models.MailFolderable], error) {
	return msgraphcore.NewPageIterator[models.MailFolderable](
		resp,
		adapter,
		models.CreateMailFolderCollectionResponseFromDiscriminatorValue,
	)
}

func iterate(iterator *msgraphcore.PageIterator[models.MailFolderable]) error {
	if err := iterator.Iterate(ctx, func(pageItem models.MailFolderable) bool {
		// Handle mail folder

		return true
	}); err != nil {
		// handle error

		return err
	}
	return nil
}
```

The `deltaLink` is always nil though.  

It doesnt work even if I use a `ItemMailFoldersDeltaRequestBuilder`:

```go
config := &users.ItemMailFoldersDeltaRequestBuilderGetRequestConfiguration{
	QueryParameters: &users.ItemMailFoldersDeltaRequestBuilderGetQueryParameters{},
}

req := me.clients.MicrosoftGraphAPI.Client().Users().ByUserId(userID).MailFolders().Delta()

resp, err := req.Get(ctx, config)
if err != nil {
	return err
}
```

But the `deltaLink` is still nil.  
It is not `nil` when I don't use the iterator (and I use the `nextLink` myself).

**Reply by maxklav** [reply]
For my previous message, should I create a bug ticket or am I not using it correctly ? :)

**Reply by baywet** [reply]
it looks correct to me but I might be missing something @rkodev do you mind helping here please?

**Reply by maxklav** [reply]
Looks like [Mail folders](https://learn.microsoft.com/en-us/graph/delta-query-overview#supported-resources) are supported for delta queries, so it should work

**Reply by luigi-fast** [reply]
Hello there,
I've tried to debug sdk code to figure out what's wrong with the delta link.  I've found that each response received by the iterator is of type *users.ItemMailFoldersDeltaResponse and the last page (the one containing the delta link) is of type *models.BaseCollectionPaginationCountResponse.  In the convertToPage method of page_iterator.go the response is casted to PageWithOdataDeltaLink but because the last page response is not of *users.ItemMailFoldersDeltaResponse type it fails and no deltaLink is returned by the iterator.

This is the first iteration:
![image](https://github.com/microsoftgraph/msgraph-sdk-go/assets/57677312/a00f9ac5-0a86-4dbc-9002-7355418795ee)

 I tried to hack the convertToPage method like this and it works perfectly.
 
 This is the last iteration:
![image](https://github.com/microsoftgraph/msgraph-sdk-go/assets/57677312/d77ef4cc-d8c4-4bd3-9326-5a908cd31523)

I've found another bug in base_collection_pagination_count_response.go, the method GetOdataNextLink look at the first level of the backingstore when it should look into backingstore/additionalData:
![image](https://github.com/microsoftgraph/msgraph-sdk-go/assets/57677312/0376d6c4-8914-4e52-89d6-9423f8b2ab8e)

I hope this will help to fix the bug

**Reply by baywet** [reply, marked as answer]
I think there are multiple confusions in the last code example you've provided.

First off
```go
req := client.Users().ByUserId(userID).MailFolders()
```
is not a delta endpoint but just the regular list endpoint, it will never return a delta link and should be instead

```go
req := client.Users().ByUserId(userID).MailFolders().Delta()
```
(the request configuration and query parameters need to be updated accordingly)

Then, it's not `models.CreateMailFolderCollectionResponseFromDiscriminatorValue` that should be used for deserialization but `users.CreateItemMailFoldersDeltaResponseFromDiscriminatorValue` https://github.com/microsoftgraph/msgraph-sdk-go/blob/3280bd480cad4749ddd7c00a6310f74c3b3ce529/users/item_mail_folders_delta_response.go#L20C6-L20C62

Then it's not `users.NewItemMailFoldersRequestBuilder` that should be used but `users.NewItemMailFoldersDeltaRequestBuilder` https://github.com/microsoftgraph/msgraph-sdk-go/blob/3280bd480cad4749ddd7c00a6310f74c3b3ce529/users/item_mail_folders_delta_request_builder.go#L47

This should solve the overall deserialization issues.

Lastly, I'd really like @rkodev to take over this discussion.

**Reply by luigi-fast** [reply]
Thanks a lot @baywet.  The use of models.CreateMailFolderCollectionResponseFromDiscriminatorValue instead of users.CreateItemMailFoldersDeltaResponseFromDiscriminatorValue was the root of the problem :) 
So it solves the issue

**Reply by andrewbcohere** [reply]
I just wanted to mention that I came across this issue, and to me it is not very intuitive, for example for iterating over drive items, it seems that instead of using `models.CreateDriveItemCollectionResponseFromDiscriminatorValue` you have to use `drives.CreateItemItemsItemDeltaResponseFromDiscriminatorValue` from `"github.com/microsoftgraph/msgraph-sdk-go/drives"`. The name is totally different (ItemItemsItem?) and I don't see in the documentation where it says this, that you may not be able to use the models package but instead have to go into another package and find the appropriate function there. Maybe I missed it, but just adding my experience here.

**Comment by jasonjoh** [comment]
We do document this in our public docs at the link you provided!

[https://learn.microsoft.com/en-us/graph/sdks/paging?tabs=go](https://learn.microsoft.com/en-us/graph/sdks/paging?tabs=go)

**Reply by baywet** [reply]
sorry, hadn't had my morning tea yet :)
